### PR TITLE
[Delivers 164829213] (Feature) - user notifications

### PR DESCRIPTION
**Pull request by vino-jasuba**
#### Description
Currently, when new events occur in the application like the publishing of a new article, existing users on the platform have no way of knowing what happened. This feature adds notifications such that users can be made aware of these events. By default, all users are "opted in" to the notifications and have the option to opt out at any time. 

#### Type of change
- [ ] Bug fix (nonbreaking change which fixes an issue)
- [x] New feature (nonbreaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

#### How Has This Been Tested?
Please describe the tests that you ran to verify your changes
- [ ] End to end
- [ ] Integration

#### Checklist:
- [ ] My code follows the style guide for this project
- [ ] I have linted my code prior to submission
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

#### Pivotal Tracker
[164829213](https://www.pivotaltracker.com/story/show/164829216)

### CODE REVIEW

File: authors/apps/articles/tests/test_article_notifications.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+                'email': 'user1@mail.com',
+                'username': 'user1',
+                'password': 'sTr0ng%weR'
+            }
+
+        self.user = User.objects.create(**user_data)
+        self.user.save()
+
+    def test_saving_article_triggers_notification(self):
+        # setup
```
Comment (oagutu) [inline authors/apps/articles/tests/test_article_notifications.py]: Get rid of/replace this commented code

File: authors/apps/authentication/serializers.py [hunk 2]
```diff
@@ -257,3 +254,33 @@ class SetNewPasswordSerializer(serializers.Serializer):
...
+class NotificationSerializer(serializers.ModelSerializer):
+    source = UserSerializer(read_only=True)
+    source_display_name = serializers.CharField(max_length=150)
+    action = serializers.CharField(max_length=50)
+    category = serializers.CharField(max_length=50)
+    url = serializers.URLField()
+    short_description = serializers.CharField(max_length=100)
+    extra_data = serializers.JSONField(default={})
+    is_read = serializers.BooleanField(default=False)
+    create_date = serializers.DateTimeField(format="%c")
```
Comment (oagutu) [inline authors/apps/authentication/serializers.py:268]: Why are these here, aren't they set automatically in the db?Alternatively if they serve a  specific purpose, then they should be `created_at` and `updated_at` to maintain consistency.

**Review by bl4ck4ndbr0wn — CHANGES_REQUESTED** [review]
Try to resolve the Merge conflicts.

File: authors/apps/comments/views.py [hunk 3]
```diff
@@ -51,7 +52,18 @@ def post(self, request, slug):
             comment = request.data.get('comment', {})
             serializer = self.serializer_class(data=comment)
             serializer.is_valid(raise_exception=True)
-            serializer.save(author=request.user, article=article)
+            comment = serializer.save(author=request.user, article=article)
+
+            # find all user's who have favorited this article
+            users = [favorite.user_id for favorite in Favorite.objects.filter(article_id=article.id)]
+            users.append(comment.article.author)
+            print(users)
```
Comment (oagutu) [inline authors/apps/comments/views.py]: Omit this print statement

File: authors/apps/core/notifications.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
+from django.template.loader import render_to_string
+
+
+def notification_channels(user):
```
Comment (oagutu) [inline authors/apps/core/notifications.py:32]: Add docstrings to these functions as well as all the other ones missing the same

File: authors/apps/core/notifications.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    :param user: User to use when building up list
+        of allowed notification channels
+    :return: list of channels
+    """
+    channels = []
+
+    if user.email_notifications:
+        channels.append('email')
+
+    if user.in_app_notifications:
```
Comment (oagutu) [inline authors/apps/core/notifications.py:43]: This block plays no purpose if you have a pass statement in it

File: authors/apps/authentication/serializers.py [hunk 6]
```diff
@@ -1,10 +1,12 @@
 import re
 
-from django.contrib.auth import authenticate
 from django.core.validators import RegexValidator, ValidationError
-from rest_framework import serializers
+from authors.apps.profiles.serializers import GetProfileSerializer
+from .models import User, PasswordReset, UserNotification
 from rest_framework.validators import UniqueValidator
-
+from notifications.models import Notification
```
Comment (oagutu) [inline authors/apps/authentication/serializers.py:7]: Is this the correct path?

**Review by oagutu — COMMENTED** [review]
Please check on your imports, along with the issues raised

File: authors/apps/authentication/serializers.py (hunk 6 above)
Comment (vino-jasuba) [reply authors/apps/authentication/serializers.py:7]: Yes ... I think what might be missing from your setup to make this not work is the package `django-notifs`. I haven't been able to figure out why but pipenv always seems to have a new problem. So I manually installed the packages locally using pip. I'll commit the requirements.txt file along with the requested changes.

File: authors/apps/authentication/serializers.py (hunk 2 above)
Comment (vino-jasuba) [reply authors/apps/authentication/serializers.py:268]: This serializer is just meant to be used when deserializing `Notification` objects when fetching them via the `api/me/notifications` endpoint

File: authors/apps/comments/views.py (hunk 3 above)
Comment (vino-jasuba) [reply authors/apps/comments/views.py]: 👍

File: authors/apps/core/notifications.py (hunk 5 above)
Comment (vino-jasuba) [reply authors/apps/core/notifications.py:43]: Was meant to be instructional for if we decide to use realtime in app notifications.

File: authors/celery.py [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+app = Celery('authors')
+
+app.config_from_object('django.conf:settings', namespace='CELERY')
+
+app.autodiscover_tasks()
+
+
+@app.task(bind=True)
+def debug_task(self):
+    print('Request: {0!r}'.format(self.request))
```
Comment (ogiste) [inline authors/celery.py]: There is a print statement in this file we need to omit.

**Review by ogiste — COMMENTED** [review]
We need to fix the merge conflicts and failing builds as well.

File: authors/apps/authentication/serializers.py (hunk 6 above)
Comment (oagutu) [reply authors/apps/authentication/serializers.py:7]: Our `Pipfile.lock` may be corrupted, you may have to delete it; and `pipenv  install` the new packages so a new piffle.lock can be created as well as update the `Pipfile`
